### PR TITLE
[ROCm][MLA] Replace torch.cat in sparse-MLA forward_mqa with fused concat_mla_q

### PR DESCRIPTION
**Pull request by maeehart**
## Summary

In the ROCm sparse-MLA backend, `forward_mqa` builds `q = torch.cat([ql_nope, q_pe], dim=-1)` once per layer per forward pass. On DeepSeek-V3.2 (61 sparse-MLA layers on the decode path) this adds up to **61 fresh tensor allocations and 61 `aten::CatArrayBatchedCopy` kernel launches per decode step**, ≈ 360 µs at MC=4 (≈ 4% of a decode step measured at TP=4, ISL=1000, OSL=100 on MI355X).

The fused `ops.concat_mla_q(ql_nope, q_pe, q_out)` already exists in `vllm/_custom_ops.py` and is already used by `FlashMLASparseImpl.forward_mqa` at `vllm/v1/attention/backends/mla/flashmla_sparse.py:1028`. This PR plugs it into `ROCMAiterMLASparseImpl` against a preallocated workspace buffer sized `(max_num_batched_tokens, num_heads, head_size)`, so:

- No per-layer allocation on the decode hot path.
- Output address is stable across CUDA graph capture/replay (same pattern `FlashMLASparseImpl` uses today).
- One fused kernel instead of one ATen cat.

The patch is numerically identical to the `torch.cat` it replaces (no change in GSM8K, see below).

The workspace buffer is allocated using `vllm_config.model_config.dtype` so both bf16 and fp16 paths advertised by `ROCMAiterMLASparseBackend.supported_dtypes` get matching-dtype storage (per @gemini-code-assist review).

## Performance

Setup: DeepSeek-V3.2 (sparse-FP8) on 4× MI355X, TP=4, MC=4, ISL=1000, OSL=100, 32 prompts per run, 4 runs per build (seeds 42/42/7/13), deterministic integer-prompt feed.

Build matrix (everything is the published nightly image except the file in this PR):

- **Baseline**: `vllm/vllm-openai-rocm:nightly-32b7177909d1c9928bcedd81de7de5a1fa21d2b3` (upstream `0.21.1rc1.dev28+g32b717790`, file `vllm/v1/attention/backends/mla/rocm_aiter_mla_sparse.py` matches `main` at the time of writing).
- **main + this PR**: same image, with this PR's `rocm_aiter_mla_sparse.py` bind-mounted on top.

Both builds run with the AITER `module_custom_all_reduce.so` from the upcoming AITER v0.14 (pinned to ROCm/aiter@`c71075ced`, "Fix 2-stage fused_allreduce_rmsnorm memory ordering", #2890) bind-mounted on top of the image's AITER 0.13. This is unrelated to the change in this PR — it is needed to keep the GSM8K accuracy comparison meaningful and is symmetric across baseline and patched (so it cannot account for the delta below).

| Build | Mean TPOT (ms) | Median TPOT (ms) | P99 TPOT (ms) | Output tok/s |
|---|---|---|---|---|
| Baseline (n=4) | 17.54 | 17.08 | 21.45 | 183.5 |
| **main + this PR** (n=4) | **17.09** | **16.92** | **18.57** | **194.1** |
| delta | **-0.46 ms (-2.6%)** | **-0.16 ms (-0.9%)** | **-2.88 ms (-13.4%)** | **+5.8%** |

Run-to-run noise on a single node is ≈ ±0.4 ms mean TPOT; the patched run is below the baseline range for every seed. The P99 improvement is the strongest signal — eliminating the per-layer ATen allocation removes a recurring tail-latency contributor.

Server:

```
vllm serve deepseek-ai/DeepSeek-V3.2 \
  --tensor-parallel-size 4 --trust-remote-code \
  --max-model-len 4096 --no-enable-prefix-caching
```

Bench:

```
vllm bench serve --dataset-name random \
  --random-input-len 1000 --random-output-len 100 \
  --random-prompt-as-token-ids \
  --num-prompts 32 --max-concurrency 4 \
  --ignore-eos --seed {42,7,13}
```

## Accuracy

GSM8K, n=200, num_fewshot=5, `--gen_kwargs temperature=0,seed=42` (greedy), evaluated against the raw `/v1/completions` endpoint (matches `.buildkite/lm-eval-harness/run-lm-eval-gsm-vllm-baseline.sh`), `transformers==4.57.1`.

| Build | flexible-extract | strict-match |
|---|---|---|
| Baseline | 0.945 ± 0.016 | 0.945 ± 0.016 |
| main + this PR | 0.945 ± 0.016 | 0.945 ± 0.016 |

Identical accuracy; expected, since the patch is numerically identical to the existing `torch.cat`.

## Single-prompt sanity (patched server, greedy)

Spot-checks against `/v1/chat/completions` with `temperature=0`:

| Prompt | Output |
|---|---|
| `What is 17 times 23? Return only the integer.` | `\( 17 \times 23 = 391 \)` |
| `What is the capital of France? Answer with the city name only.` | `Paris` |
| `What is the largest planet in our solar system? Reply with one word.` | `Jupiter` |
| `A train leaves at 3:00 PM and arrives 2.5 hours later. What time does it arrive in HH:MM 24-hour format? Reply with just HH:MM.` | `17:30` |

All four expected answers; no spurious continuations, no language drift.

## Static checks

- `python tools/pre_commit/mypy.py 1 "3.10"` — clean
- mypy for Python 3.11, 3.12, 3.13 (full CI matrix) — clean
- `ruff check vllm/v1/attention/backends/mla/rocm_aiter_mla_sparse.py` — All checks passed
- `ruff format --check vllm/v1/attention/backends/mla/rocm_aiter_mla_sparse.py` — already formatted

## Cross-reference

Mirrors the workspace-buffer + `ops.concat_mla_q` pattern that `FlashMLASparseImpl.forward_mqa` already uses at `vllm/v1/attention/backends/mla/flashmla_sparse.py:1028`.

cc @tjtanaa

### CODE REVIEW

**Comment by maeehart** [comment]
@tjtanaa would you mind taking a look? This mirrors the workspace-buffer + `concat_mla_q` pattern `FlashMLASparseImpl` already uses, so the change should be very surgical for review. Bench, GSM8K, and single-prompt sanity numbers are in the PR description. Static checks (mypy 3.10-3.13, ruff) are clean.

File: vllm/v1/attention/backends/mla/rocm_aiter_mla_sparse.py [hunk 1]
```diff
@@ -612,6 +613,21 @@ def __init__(
...
+        # graph capture/replay. Mirrors the pattern already used by
+        # `FlashMLASparseImpl` (see `flashmla_sparse.py`).
+        vllm_config = get_current_vllm_config()
+        max_tokens = vllm_config.scheduler_config.max_num_batched_tokens
+        # `head_size` here already equals `kv_lora_rank + qk_rope_head_dim`,
+        # i.e. the post-concat last-dim that the MLA kernels consume.
+        q_concat_shape = (max_tokens, num_heads, head_size)
+        (self.q_concat_buffer,) = current_workspace_manager().get_simultaneous(
+            (q_concat_shape, torch.bfloat16),
+        )
```
[the remark this answers is not in the document]
Comment (maeehart) [reply vllm/v1/attention/backends/mla/rocm_aiter_mla_sparse.py:653]: Good catch, fixed in the latest force-push: the buffer is now allocated with `vllm_config.model_config.dtype` so both bf16 and fp16 (the two dtypes `ROCMAiterMLASparseBackend.supported_dtypes` advertises) get matching-dtype storage. Also expanded the comment block above the allocation to document the constraint so future readers don't have to re-derive it.

```python
        q_concat_shape = (max_tokens, num_heads, head_size)
        (self.q_concat_buffer,) = current_workspace_manager().get_simultaneous(
            (q_concat_shape, vllm_config.model_config.dtype),
        )
```

GSM8K and bench on bf16 (DSV3.2) match baseline; tested on MI355X TP=4 with the patched build. No fp16-capable sparse-MLA model handy to bench, but the only thing changing for fp16 users is the dtype tag on a single workspace allocation.

File: vllm/v1/attention/backends/mla/rocm_aiter_mla_sparse.py [hunk 2]
```diff
@@ -612,6 +613,25 @@ def __init__(
         assert indexer is not None
         self.topk_indices_buffer: torch.Tensor | None = indexer.topk_indices_buffer
 
+        # Workspace buffer for `concat_mla_q`, which replaces the
```
Comment (Rohan138) [inline vllm/v1/attention/backends/mla/rocm_aiter_mla_sparse.py]: LGTM, but can we remove the ~18 lines of comments? The only comment in  `flashmla_sparse.py` is `# Concatenate q if it's a tuple (ql_nope, q_pe)`

**Review by Rohan138 — APPROVED** [review]

File: vllm/v1/attention/backends/mla/rocm_aiter_mla_sparse.py (hunk 2 above)
Comment (maeehart) [reply vllm/v1/attention/backends/mla/rocm_aiter_mla_sparse.py]: Done in 0a04d40 (force-push). Dropped both comment blocks at the buffer allocation and trimmed the `forward_mqa` comment back to the same one-liner `flashmla_sparse.py` uses. The performance / decode-time motivation is now only in the PR body and commit message.

```python
        vllm_config = get_current_vllm_config()
        max_tokens = vllm_config.scheduler_config.max_num_batched_tokens
        q_concat_shape = (max_tokens, num_heads, head_size)
        (self.q_concat_buffer,) = current_workspace_manager().get_simultaneous(
            (q_concat_shape, vllm_config.model_config.dtype),
        )
```

```python
        # Concatenate q if it's a tuple (ql_nope, q_pe)
        if isinstance(q, tuple):
            ql_nope, q_pe = q
            q = self.q_concat_buffer[: ql_nope.shape[0]]
            ops.concat_mla_q(ql_nope, q_pe, q)
```

**Comment by maeehart** [comment]
Heads-up: the force-push from `106a28763b` → `0a04d40fc7` stale-dismissed the approval. The only diff between the two revisions is the comment removal you requested in [#discussion_r3270721693](https://github.com/vllm-project/vllm/pull/42838#discussion_r3270721693) — no code changes — so this should be a quick re-stamp.

@Rohan138 could you re-approve `0a04d40fc7` when you get a chance? Thanks!

**Comment by maeehart** [comment]
Could I get this merged @dllehr-amd and @Rohan138.

**Comment by maeehart** [comment]
### Rebase on current main

Rebased this PR onto current `upstream/main` (HEAD `5c1aec3dc0`). The branch is now a single commit `ebbf254bac` with the same 12-line diff as before (workspace-buffer allocation + `concat_mla_q` call in `forward_mqa`).

This refresh should re-trigger CI on the latest base. The earlier `pre-run-check` failure on the status rollup is from a run that pre-dated the `ready` label and a fresh run picks it up correctly. The `docs/readthedocs.org:vllm` failure on the rollup tracks a problem that is also showing up on several other vLLM PRs right now, so it does not look related to this change.

@Rohan138 your earlier LGTM was stale-dismissed by the previous force-push. Both your "drop the comment block" feedback and the gemini "buffer dtype must follow `model_config.dtype`" comment were addressed in the rebased commit. Could I get a re-review?

@dllehr-amd flagging for visibility.

**Comment by Rohan138** [comment]
@maeehart can you delete the cuda bf16/fp16 headers from https://github.com/vllm-project/vllm/blob/main/csrc/concat_mla_q.cuh#L4...L5? `cuda_vec_utils.cuh` already conditionally includes either the cuda or hip bf16/fp16 header files respectively. Else LGTM, thanks for the PR!

Note: After discussing with @dllehr-amd and Claude, the rest of the kernel is platform-agnostic. We end up running 2 heads per 64-thread wavefront on rocm instead of 1 head per 32-thread warp on cuda, but since we have e.g. q_pe = [num_tokens, num_heads, rope_head_dim = 64]; 64 * 2 bytes per bf16 value = 128 bytes; with 4-byte load/store instructions we have 128/4 = 32 parallel loads per head as the optimal "per-head parallelism" anyway. This is why can reuse the same kernel with 32-thread based indexing, instead of modifying it for ROCm's `wave64` arch:
```
  const int flat_warp_id = (blockIdx.x * blockDim.x + threadIdx.x) >> 5;
  const int token_id = flat_warp_id / num_heads;
  const int head_id = flat_warp_id % num_heads;
  const int lane_id = threadIdx.x & 31;
```

So for e.g. on ROCm, in the first wavegroup, we will have both `flat_warp_id 0` for `threadIdx 0...31` and `flat_warp_id 1` for `threadIdx 32...63`, while `lane_id = [0, 1, ..., 30, 31, 0, 1, ..., 30, 31]`. The names for `flat_warp_id` and `lane_id` are being abused a bit here, but the kernel ends up being correct+performant.

**Review by dllehr-amd — APPROVED** [review]
Thanks @maeehart   I appreciate the PR!

**Comment by dllehr-amd** [comment]
Updating branch.  Lets see if CI comes back clean after this

**Comment by dllehr-amd** [comment]
@maeehart  Can you address @Rohan138's query?  In the kernel we are double importing cuda_bf16.h as the vec_utils will do it too, but actually guard on the ROCm vs CUDA.  Also do you mind rebasing this.  The file structure has changed ever so slightly.  Otherwise I think it's good to go!

**Comment by frida-andersson** [comment]
@dllehr-amd @Rohan138 force-pushed `47ecd26e8` — linear 2 commits on top of today's main.

- Rebased.
- @Rohan138's `<cuda_bf16.h>`/`<cuda_fp16.h>` removal applied at the new `csrc/libtorch_stable/concat_mla_q.cuh` path as a separate commit.

Re-checked post-rebase: ruff clean; profile confirms no `CatArrayBatchedCopy` in the sparse-MLA decode hot path. Python hunk is byte-identical to the originally-reviewed version, so the GSM8K + bench numbers in the PR description carry forward unchanged. Ready for a re-stamp + merge whenever you have a moment. Thanks!

**Review by tjtanaa — APPROVED** [review]
LTGM, #include <cuda_bf16.h> and #include <cuda_fp16.h> are both included in "cuda_vec_utils.cuh"
